sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/CachingBeamFnStateClient.java

```diff
@@ -0,0 +1,167 @@
+            StateCacheKey.create(cacheToken, request.getGet().getContinuationToken());
+        Map<StateCacheKey, StateGetResponse> stateKeyMap = stateCache.getUnchecked(stateKey);
+        cachedPage = stateKeyMap.get(cacheKey);
+
+        if (cachedPage == null) {
+          beamFnStateClient.handle(requestBuilder, response);
+          CompletableFuture<Void> callback =
+              response.thenAccept(
+                  stateResponse -> {
+                    stateCache.getUnchecked(stateKey).put(cacheKey, stateResponse.getGet());
+                  });
+
+          callback.getNow(null);
+        } else {
+          response.complete(
+              StateResponse.newBuilder().setId(requestBuilder.getId()).setGet(cachedPage).build());
+        }
+
+        return;
+
+      case APPEND:
+        // Treat appends as normal for now and do not cache
+        beamFnStateClient.handle(requestBuilder, response);
+        return;
+
+      case CLEAR:
+        Map<StateCacheKey, StateGetResponse> clearedData = new HashMap<>();
+        StateCacheKey newKey = StateCacheKey.create(cacheToken, ByteString.EMPTY);
+        clearedData.put(newKey, StateGetResponse.getDefaultInstance());
+        stateCache.put(stateKey, clearedData);
+        beamFnStateClient.handle(requestBuilder, response);
+        return;
+
+      default:
+        throw new IllegalStateException(
+            String.format("Unknown request type %s", request.getRequestCase()));
+    }
+  }
+
+  private ByteString getCacheToken(BeamFnApi.StateKey stateKey) {
+    // type: (beam_fn_api_pb2.StateKey) -> Optional[bytes]
+    if (stateKey.hasBagUserState()) {
+      return userStateToken;
+    } else if (stateKey.hasRunner()) {
+      // TODO: Support runner state key caching
+      return ByteString.EMPTY;
+    } else {
+      CacheToken.SideInput.Builder sideInputBuilder = CacheToken.SideInput.newBuilder();
+      if (stateKey.hasIterableSideInput()) {
+        IterableSideInput iterableSideInput = stateKey.getIterableSideInput();
+        sideInputBuilder
+            .setTransformId(iterableSideInput.getTransformId())
+            .setSideInputId(iterableSideInput.getSideInputId());
+      } else if (stateKey.hasMultimapSideInput()) {
+        MultimapSideInput multimapSideInput = stateKey.getMultimapSideInput();
+        sideInputBuilder
+            .setTransformId(multimapSideInput.getTransformId())
+            .setSideInputId(multimapSideInput.getSideInputId());
+      } else if (stateKey.hasMultimapKeysSideInput()) {
+        MultimapKeysSideInput multimapKeysSideInput = stateKey.getMultimapKeysSideInput();
+        sideInputBuilder
+            .setTransformId(multimapKeysSideInput.getTransformId())
+            .setSideInputId(multimapKeysSideInput.getSideInputId());
+      }
+      return sideInputCacheTokens.getOrDefault(sideInputBuilder.build(), ByteString.EMPTY);
+    }
+  }
+
+  @AutoValue
+  public abstract static class StateCacheKey {
```

Comment:
Add comments please. Note: every class, public method, and tricky bits of code should have comments. There are some exceptions of course, like `getters/setters` or trivial self explanatory methods.

If you're curious about learning more, you can refer to Google's public style guide for some good practices: https://google.github.io/styleguide/javaguide.html